### DISCUSSION TITLE
SIP TCP Request/Response use existing TCP connection rather than, resolving DNS and creating a new TCP Request

### DISCUSSION THREAD

**Discussion by aankur** [resiprocate/resiprocate#373] [Q&A] [answered]
Hi,

When an incoming call comes in and we respond to it asynchronously.
like INFO command or Bye Command, is send as a new connection by resolving DNS, to the contact of the incoming call, that might be in a separate ip network,

The DNS Call might not always succeed,
But The original connection that the call was made from still exists and is connected,
how do we make resiprocate respond to the original connection rather than try to respond with a new SIP TCP connection?

**Comment by sgodin** [comment, marked as answer]
In RFC3261 for TCP/TLS - the actual SIP response (ie: 200) to an incoming SIP request always travels down the same connection the request was received on.  However when sending a new request back to the sender (note: you are calling this a response above), the default behavior is to form a new TCP (or TLS) connection in the other direction.  The IETF created RFC5626 to address this issue, since it is not always possible to establish a connection in the other direction when firewalls are involved.  https://datatracker.ietf.org/doc/html/rfc5626

Resip specific info here:  https://github.com/resiprocate/resiprocate/wiki/DUM-Client-Outbound-Support

If you are using repro it has a flow token feature and RFC5626 support to help with this.
https://github.com/resiprocate/resiprocate/wiki/Using-Repro
In particular:
DisableOutbound = false
EnableFlowTokens = true

WebRTC needs this behavior, so this link is also good (for repro):
http://www.sipspectrum.com/blog/configuring-repro-for-webrtc

**Reply by aankur** [reply]
BYE does not work, it always tries t create a new connection to send BYE

**Comment by aankur** [comment]
Thanks, will test and respond

**Comment by sgodin** [comment]
If you can describe the problem in more detail, then I may be able to help.  A description of whether you are using DUM or repro is needed. 
Also resip logs and/or pcap traces would be very useful.

**Reply by aankur** [reply]
i am using DUM, i will collect and add Logs

**Reply by aankur** [reply]
i am setting the following options
```
// RFC5626 support
  resip::InteropHelper::setRRTokenHackEnabled(true);
  resip::InteropHelper::setOutboundSupported(true);
  resip::InteropHelper::setOutboundVersion(5626);

  // RFC5626 support
  mProfile->addSupportedOptionTag(resip::Token(resip::Symbols::Outbound));  // RFC 5626 - outbound
  mProfile->addSupportedOptionTag(resip::Token(resip::Symbols::Path));      // RFC 3327 - path
  mProfile->setExpressOutboundAsRouteSetEnabled(true);
  mProfile->setKeepAliveTimeForStream(60);
  mProfile->setInstanceId(resip::Data("inst-1"));  // See RFC5626 section 4.1
  mProfile->setRegId(1);
  mProfile->clientOutboundEnabled() = true;

```

Logs
[sipcall3.log](https://github.com/user-attachments/files/17323868/sipcall3.log)

As you can see INFO are send via a new connection
and BYE is timing out

**Comment by sgodin** [comment]
In order for RFC5626 to work you need to register correctly.  In your logs the registration is failing...

DEBUG | 20241010-070304.501 | SIPServer | RESIP | 140286356457216 | ParseBuffer.cxx:983 | ParseBuffer.cxx:747, Parse failed Expected a digit in context: Request Line
REGISTER sip:[IP]:[Port] SIP/2.0
                  ^

DEBUG | 20241010-070304.501 | SIPServer | RESIP | 140286356457216 | BaseException.cxx:21 | BaseException at ParseBuffer.cxx:747 ParseBuffer.cxx:747, Parse failed Expected a digit in context: Request Line
REGISTER sip:[IP]:[Port] SIP/2.0
                  ^

INFO | 20241010-070304.501 | SIPServer | RESIP | 140286356457216 | Helper.cxx:1770 | Illegal request line
INFO | 20241010-070304.501 | SIPServer | RESIP:TRANSPORT | 140286356457216 | Transport.cxx:411 | Message Failed basicCheck :SipReq:  REGISTER IP tid=-d87543-790058255-1--d87543- cseq=1 REGISTER contact=%5BUser%5D@Local / 1 from(wire) tlsd=sip.test.example


Also keep in mind if you have more than one UserProfile, you also need to assign any new inbound messages to the correct UserProfile, via:  selectUASUserProfile:  https://github.com/resiprocate/resiprocate/wiki/DUM-Configuration-and-Profiles


Here's my summary of what I see in the logs:

Because you didn't register, then RFC5626 is not getting used.  It's also not if you are using the correct UserProfile when receiving the inbound request...


Inbound invite from:  117.216.47.145:65210

INVITE sip:5790109175@sip.test.example;transport=tcp SIP/2.0
Via: SIP/2.0/TCP 192.168.7.246:5060;branch=z9hG4bK2E7B0D3E19b2ecb0f607a8c0
Max-Forwards: 70
Contact: <sip:user4433@192.168.7.246:5060;transport=tcp>;+sip.instance="<urn:uuid:7f40ee50-8ff2-103d-8122-4fc35fedc55e>"
To: <sip:5790109175@sip.test.example>
From: <sip:user4433@192.168.7.246>;tag=15DE2970-9459F7A3
Call-ID: d846701-89df5bac-7ab6cbff@192.168.7.246

Outbound 180 and 200 route back on same connection as all responses do

Inbound ACK received from: 117.216.47.145:65210

Inbound reinvite received from: 117.216.47.145:65210

Outbound INFO created, and will send to address is Contact header:  192.168.7.246:5060

It created a new outbound connection for this:
INFO | 20241010-070311.881 | SIPServer | RESIP:TRANSPORT | 140286356457216 | Connection.cxx:45 | Connection::Connection: new connection created to who: [ V4 192.168.7.246:5060 TCP targetDomain=192.168.7.246 flowKey=21 transportKey=1 ], is server = 0


No response was ever received for the INFO - do stack generated a 408 for DUM

Outbound BYE created and sent to the exact same address as INFO:

DEBUG | 20241010-070412.177 | SIPServer | RESIP:TRANSPORT | 140286356457216 | TransportSelector.cxx:1335 | Transmitting to [ V4 192.168.7.246:5060 TCP targetDomain=192.168.7.246 transportKey=1 ] tlsDomain= via [ V4 172.31.43.144:5060 TCP transportKey=1 ]

BYE sip:user4433@192.168.7.246:5060;transport=tcp SIP/2.0

There was also no response for it, and the stack generated a 408 for DUM

**Reply by aankur** [reply]
Hi,
this is a standalone instance, no register is expected of it,
it is supposed to accept calls and respond to it.
180 and 200 were routed back to the same connection.

why was outbound INFO sent to Contact header of a NAT-ed endpoint and  not reuse the same connection. like 180 and 200.

also how to ensure outbound INFO is sent via the same connection connected to 117.216.47.145:65210 rather than a new connection via the contact header, as the endpoint is behind NAT
